proposals/0000-typed-metadata.md

```diff
@@ -0,0 +1,528 @@
+
+// error: Type `haxe.Exception` is not valid argument type for metadata function.
+@:metadata function invalidType(o: haxe.Exception): Any;
+```
+
+&nbsp;
+&nbsp;
+
+## Haxe API Changes
+
+A new optional field should be added to `haxe.macro.Expr.MetadataEntry`.
+
+If this metadata entry is typed, then `field` will contain a reference to the `ClassField` of the metadata function.
+```haxe
+// Unresolved question
+// Would it be possible to use Ref<Type.ClassField> instead?
+var ?field: Expr.Field;
+```
+
+### Reading Arguments
+
+There needs to be a mechanism for reading metadata arguments. To provide this, new class should be added: `haxe.macro.MetadataEntryTools`.
+
+This class provides static extension functions for `MetadataEntry` for reading arguments. Each function attempts to read an argument for a specific type. For example, for the `Int` argument type, there should be a `getInt(index: Int)` function that looks like this:
+
+```haxe
+static function getInt(entry: MetadataEntry, index: Int): Null<Int> {
+	return if(entry.params != null && index < entry.params.length) {
+		switch(entry.params[index].expr) {
+			case EConst(CInt(v)): Std.parseInt(v);
+			case _: null;
+		}
+	} else {
+		null;
+	}
+}
+```
+
+There should be one function for every possible argument type.
+
+These functions should not throw any errors; that is the job of the Haxe compiler on typed metadata. Instead `null` is returned if the argument doesn't exist or doesn't match the desired type. Technically, these could also be used on untyped metadata.
+
+### Function Type Struct
+
+The following anonymous structure should be added to the `haxe/macro/Expr.hx` module:
+```haxe
+typedef FunctionPath = {
+	> TypePath,
+	functionName: String;
+};
+```
+
+This is a structure for storing type paths to functions. It is used as an argument type for metadata. Long story short, it allows for type paths that end with a lowercase identifier (`myFunc`, `Module.Sub.myFunc`).
+
+Technically, function path data _could_ be stored in `TypePath`, but that's not preferable.
+
+&nbsp;
+&nbsp;
+
+## New Metadata
+
+There needs to be a way for metadata functions to configure a couple options:
+ * Can it be used multiple times on the same subject?
+ * Is it compile-time only? Or should it exist as rtti.
+ * Is it restricted to one or more platforms?
+ * Does it require another metadata to function?
+
+While likely not set in stone, this proposal recommends adding the following metadata to be used in combination with `@:metadata` to configure these options:
+```haxe
+/**
+	Use on a metadata function. That meta will only exist at
+	compile-time (will not generate any rtti data).
+
+	Unresolved question: Should this force this meta to
+	be prefixed with a colon? There needs to be a way to signify
+	that when typing built-in Haxe metadata in the future.
+**/
+@:metadata
+@metadataCompileOnly
+@metadataRequire(@:metadata _)
+function metadataCompileOnly(): haxe.macro.Expr.Function;
```

Comment:
An alternative would be to use `@metadata function ...` to define a runtime metadata (`@foo`), and `@:metadata function ...` for a compile-time one (`@:foo`). Requiring to add `@metadataCompileOnly` to all compile-time metadata sounds very annoying (also, we have namespaces).

Drawback is that it's easy to make a mistake between `@:metadata function` and `@metadata function` but that should be caught very easily at use site.

Reply:
> also, we have namespaces

Btw, this seems like a limitation of this proposal to not handle metadata namespaces, which is unfortunate.

Reply:
> (edit: I guess it's in concurrence with fully qualified path there..)

Oh yeah, sorry if I didn't make it as clear in the proposal but you import like normal functions, otherwise you use full path like you would with a normal function call. I kind of touch on it a little [here](https://github.com/RobertBorghese/haxe-evolution/blob/ae1a4b9a9856e2f0653436973d60bdd64fc51cf5/proposals/0000-typed-metadata.md?plain=1#LL143C7-L143C7), demonstrating how using module-level vs static function changes how you'd write it with basic module import.

Reply:
> Drawback is that it's easy to make a mistake between `@:metadata function` and `@metadata function` but that should be caught very easily at use site.

Yeah, this was certainly a tough thing to find an answer to. Maybe typed metadata should just support being used both ways. But if there's a colon, it doesn't generate rtti? Hmmm...